---
title: "Salmonella enterica basemod analysis"
author: "Auke Haver"
date: "Januari 12, 2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
#install.packages("ggplot2")
library(ggplot2)
#install.packages("rtracklayer")
library(rtracklayer,warn.conflicts = FALSE)
#install.packages("tidyverse")
library(tidyverse,warn.conflicts = FALSE)
#install.packages('dendextend')
source("salmonella_methylation_pipeline_analysis_scripts.R")
library(kableExtra,warn.conflicts = FALSE)
library(gridExtra)
```

# Read design file, protein files and annotated virulence factor results from the gene analysis
```{r}
# read info file, only including the hybrid basemods
design <- read_tsv("../info/analysis_design.tsv",col_types="ccccc") %>%
  filter(Sequence_type == "hybrid")
# non-redundant protein database
nr_pdb <- read.csv("../protein_database/reference_files/senterica_pdb_non_redundant.csv")
# duplicate protein database
d_pdb <- read.csv("../protein_database/reference_files/senterica_pdb_duplicates.csv")
# highly similar (hs) core-genome (cg) virulence factor (vf) genes
hs_cg_vf_genes <- read_tsv("../output/gene_analysis/highly_similar_cg_vf_genes.tsv", col_types = "ciciicccccicccc")

```

# Annotate basemods (+/-8 min per sample)
```{r}
for(i in 1:nrow(design)){
  annotate_basemods(design$Sample[i],design$Reference[i])
}

```

# Match motifs to basemods (+/-3 min per sample)
```{r}
make_motif_matches(design,"../smrtlink/")
```

# Read motif TSV files into a dataframe
```{r}
# Iterate over all samples
for(i in 1:nrow(design)){
  # Initiate the dataframe from the first sample
  if(i==1){
    motif_overview_df <- 
      # Make file path
      paste0("../smrtlink/matched_motifs_",design$Sample[i],".tsv") %>%
      # Read file
      read_tsv(col_types = "cciiccnncliiicc")
  }
  # Append all other samples to the dataframe
  else{
    motif_overview_df <- 
      # Make file path
      paste0("../smrtlink/matched_motifs_",design$Sample[i],".tsv") %>%
      # Read file
      read_tsv(col_types = "cciiccnncliiicc") %>%
      # Append to dataframe
      full_join(motif_overview_df,by = colnames(motif_overview_df))
  }
}
```

# Read genes into a dataframe
```{r}
gene_annotations <- read_annotation_overview(design,"../REF/",nr_pdb)
```

# From the gene analysis, import all RM elements and information on associated motifs
```{r}
observed_motif_annotations <- "../output/gene_analysis/rm_overview_latex_table.tsv" %>% 
  read_tsv(col_types = "cccic") %>%
  pull(Motif) %>%
  unique()
```



# Correct motif annotations for possible motifs and duplicates
```{r}
# First, check if a motif is duplicates in the dataframe (for instance both CGACG and GCAGC are annotated).
annotated_motifs <- motif_overview_df$motif %>% unique()
# Make a list of unique motifs
unique_motifs <- vector(mode="list")
# Loop over all motifs and only add them to the unique list if neither the motif nor the complement is already in the list
for(i in 1:length(annotated_motifs)){
  if(!annotated_motifs[i] %in% unique_motifs & !reverse_complement(annotated_motifs[i])%in% unique_motifs){
    unique_motifs <- c(unique_motifs,annotated_motifs[i])
  }
}
unique_motifs <- unique_motifs%>% unlist()
corrected_motif_overview_df <- motif_overview_df %>%
  rowwise()%>%
  mutate(motif = ifelse(!motif %in% unique_motifs,reverse_complement(motif),motif),
         # Then, check if the motif is a match with at least one MTase.
          # If not, check if its reverse complement is and if so, change the motif to its reverse complement.
         motif = ifelse(reverse_complement(motif) %in% observed_motif_annotations,reverse_complement(motif),motif))
```


# Make an overview of all motif counts in each sample, both for all genes and only in virulence genes
```{r}
# Make a list of annotated virulence gene locustags
annotated_vf_loci <- gene_annotations %>%
  # Filter for Virulence genes
  filter(Source=="VFDB"|Source=="UniProt_VF")%>%
  # Pull column into list
  pull(Name)

# Plot all motif counts per group
corrected_motif_overview_df %>%
  rowwise() %>%
  # Check wether a gene is a virulence factor
  mutate(motif = short_motif_form(motif),
         vf_motif = ifelse(Name %in% annotated_vf_loci,TRUE,FALSE))%>%
  ungroup() %>%
  group_by(motif,sample) %>%
  # Count all occurances of a motif per sample, either in all genes or in virulence genes
  summarize(motif_count = n(),vf_motif_count =sum(vf_motif)) %>%
  group_by(sample)%>%
  # Calculate percentages of motif counts per sample
  # change motif short form
  mutate(`Annotated Database Genes` = motif_count/sum(motif_count),
         `Annotated Virulence Genes` = vf_motif_count/sum(vf_motif_count)) %>%
  pivot_longer(cols=c("Annotated Database Genes","Annotated Virulence Genes"),names_to="genes",values_to="percentage")%>%
  # Add serovar information
  left_join(design %>% select(Sample,Source),
            by=c("sample"="Sample"))%>% 
  # Plot
  ggplot(aes(x=gsub("_"," ",Source),
             y=percentage*100,
             colour=motif,
             shape=genes))+
  geom_jitter(stat="identity",size=3,width=0.2)+
  scale_colour_viridis_d()+
  theme_bw()+
  theme(text=element_text(family="Times New Roman",color="black",size=10),
        axis.text.x=element_text(angle=45,hjust=1,size=10,color="black"),
        axis.text.y = element_text(size=10,color="black"),
        legend.text = element_text(size=8),
        legend.position = "bottom",
        legend.box="vertical",
        aspect.ratio = 1.2)+
  labs(x="Serovar / Source",y="Percentage of all motifs",colour="Methylation\nMotif",shape="Location\nwithin\nGenome")
ggsave("../output/basemod_analysis/motif_distribution.png")
```

# Match all motifs to mtase genes
```{r}
# Make a dataframe
motif_mtase_match <- 
  corrected_motif_overview_df %>% 
  group_by(sample,motif)%>%
  summarize(motif_counts = n()) %>%
  left_join(design %>% select(Sample,Source),
            by=c("sample"="Sample")) %>%
  ungroup()%>%
  select(Source,motif,motif_counts) %>%
  left_join("../output/gene_analysis/rm_overview_latex_table.tsv" %>% 
              read_tsv(col_types = "cccic"),
            by=c("Source"="Sample","motif"="Motif")) %>%
  filter(!is.na(motif_counts))

# Plot
motif_mtase_match %>%
  filter(!is.na(`Putative RM element`))%>%
  complete(Source,`Putative RM element`)%>%
  group_by(`Putative RM element`) %>%
  rowwise()%>%
  mutate(motif_counts = ifelse(is.na(motif_counts),0,motif_counts)) %>%
  ggplot(aes(x=Source,y=`Putative RM element`, fill=log(motif_counts)))+
  geom_tile()+
  scale_fill_viridis_c(na.value="white")+
  theme_minimal()+
  labs(fill="Motif count")+
  theme(legend.position = "bottom",axis.text.x = element_text(angle=45,hjust=1))+
  geom_text(aes(label=motif_counts))
ggsave("../output/basemod_analysis/mtase_motif_match.png")
# input heatmap with number of possible mtase
```


```{r}
# load highly similar (hs) core genome (cg) virulence factor genes
hs_cg_vf_genes <- "../output/gene_analysis/highly_similar_cg_vf_genes.tsv" %>%
  read_tsv(col_types="ciciicccccicccc")%>%
  select(sample,Name)


# Load annotated_basemods
for(i in 1:nrow(design)){
  if(i==1){
    hs_cg_vf_basemods <- paste0("../smrtlink/annotated_motifs/motifs_", design$Sample[i],".gff") %>%
      readgff() %>% 
      filter(Name %in% hs_cg_vf_genes$Name) %>%
      left_join(hs_cg_vf_genes %>% filter(sample==design$Sample[i]),
                by="Name") %>%
      select(sample,start,Name)
  }
  else{
    hs_cg_vf_basemods <- paste0("../smrtlink/annotated_motifs/motifs_", design$Sample[i],".gff") %>%
      readgff() %>% 
      filter(Name %in% hs_cg_vf_genes$Name) %>%
      left_join(hs_cg_vf_genes%>% filter(sample==design$Sample[i]),
                by="Name") %>%
      full_join(hs_cg_vf_basemods,
                by = colnames(hs_cg_vf_basemods)) %>%
      select(sample,start,Name)
  }
}
```

# Calculate basemod differences between strains
```{r}
basemod_distances <- hs_cg_vf_basemods %>% 
  rename(location=start)%>%
  left_join(gene_annotations %>% select(sample,Name,gene,start,end),by=c("sample","Name")) %>%
  rowwise()%>%
  mutate(index = ifelse(location > end,-1,location-start),
         modified = 1) %>%
  filter(index>=0) %>%
  mutate(gene_index = paste0(gene,"_",index)) %>%
  select(sample,gene_index,modified) %>%
  complete(sample,gene_index)%>%
  mutate(modified = ifelse(is.na(modified),0,1)) %>%
  distinct()%>%
  # Pivot into wide dataframe
  pivot_wider(names_from = gene_index,values_from=modified) %>%
    # change sample names to rownames in order not to lose them later
  column_to_rownames("sample")
basemod_distances %>%
  t() %>%
  as_tibble()%>%
  mutate(position = colnames(basemod_distances))%>%
  relocate(position)%>%
  write_tsv("../output/basemod_analysis/hs_cg_vf_basemod_positions.tsv")
basemod_distances <- basemod_distances%>%
  # Turn into matrix,caluclate distance and turn back into matrix and then tibble
  as.matrix()%>%
  dist(method="manhattan") %>% 
  as.matrix()%>%
  as_tibble()

# return rowname information
row.names(basemod_distances)<-colnames(basemod_distances)

# make heatmap
basemod_distances %>% rownames_to_column("query")%>%
  pivot_longer(cols=colnames(basemod_distances),names_to="match",values_to="distances") %>%
  rowwise()%>%
  mutate(query = design$Source[design$Sample==query],
         match = design$Source[design$Sample==match]) %>%
  ggplot(aes(x=gsub("_"," ",query),y=gsub("_"," ",match),fill=distances))+
  geom_tile()+
  theme_minimal()+
  scale_fill_viridis_c()+
  geom_text(aes(label=distances))+
  theme(text=element_text(family="Times New Roman",size=15,color="black"),
    axis.text.x=element_text(color="black",angle=45,hjust=1),
        axis.text.y=element_text(color="black"),
        axis.title.x = element_text(color="black"),
        aspect.ratio = 1)+
  labs(fill="Position\nSpecific\nModification\nDifferences",y="Strain",x="Strain")
ggsave("../output/basemod_analysis/basemod_differences_heatmap.png")
```

# Calculate basemod counts per gene per strain
```{r}
hs_cg_vf_basemods %>%
  left_join(gene_annotations %>% select(sample,Name,gene),
            by=c("sample","Name")) %>%
  group_by(sample,Name,gene) %>%
  summarize(basemod_count=n())%>%
  left_join(design %>% select(Sample,Source,Serovar),
            by=c("sample"="Sample")) %>%
  group_by(sample,gene)%>%
  mutate(multiplicit = ifelse(n()>1,TRUE,FALSE),
         rownumber=as.integer(row_number())) %>%
  rowwise()%>%
  mutate(gene = ifelse(multiplicit,
                       paste0(gene,rownumber),
                       gene))%>%
  ungroup()%>%
  group_by(gene)%>%
  mutate(adjusted_basemod_count = basemod_count/mean(basemod_count))%>%
  ggplot(aes(x=gene,y=gsub("_"," ",Source),fill=log10(basemod_count)))+
  geom_tile()+
  scale_fill_viridis_c()+
  labs(fill="Log\nBasemod\nCount",y="Strain", x="Shared Virulence Factor")+
  theme(text=element_text(family="Times New Roman",size=8),
        axis.text.x=element_text(color="black",angle=90,hjust=1,size=8),
        axis.text.y=element_text(color="black",size=8),
        axis.title.x=element_text(size=8,hjust=.6),
        axis.title.y = element_text(size=8),
        legend.title = element_text(size=8),
        aspect.ratio = 0.3)
ggsave("../output/basemod_analysis/hs_cg_vf_heatmap.png")
 
```


