---
title: "Salmonella enterica gene analysis"
author: "Auke Haver"
date: "December 28, 2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
#install.packages("ggplot2")
library(ggplot2)
#install.packages("rtracklayer")
library(rtracklayer,warn.conflicts = FALSE)
#install.packages("tidyverse")
library(tidyverse,warn.conflicts = FALSE)
#install.packages('dendextend')
source("salmonella_methylation_pipeline_analysis_scripts.R")
library(kableExtra,warn.conflicts = FALSE)
#install.packages("VennDiagram")
library(VennDiagram)
library(xtable)

```

```{r}
# Design used for analysis
analysis_design <- 
  "../info/analysis_design.tsv" %>% 
  read_tsv(col_types="ccccc")
# Non-redundant protein database
nr_pdb <- 
  "../protein_database/reference_files/senterica_pdb_non_redundant.csv" %>%
  read.csv()
# Duplicate PDB
d_pdb <- "../protein_database/reference_files/senterica_pdb_duplicates.csv" %>%
  read.csv()
# Virulence genes
virulence_factors <- full_join(
  "../protein_database/download/Uniprot_VF_list.csv" %>% read_csv(col_types="ccc"),
  "../protein_database/download/VFDB_VF_list.csv" %>% read_csv(col_types="cc"),
  by=c("Locus","Gene")) %>%
  select(-Uniprot)
```

# Read genes into a dataframe
```{r}
### Function to read annotated CDS into dataframe and combine with protein database 
annotated_genes <- read_annotation_overview(analysis_design,"../REF/",nr_pdb) %>%
  rowwise()%>%
  mutate(Locus = ifelse(Source=="UniProtKB" & is.na(Locus),
                        gsub("_.$","",gene),
                        Locus),
         Locus = ifelse(Source=="cgMLST" & is.na(Locus),
                        gsub("_.$","",gene),
                        Locus))
### select all annotated virulence genes
annotated_virulence_genes <- 
  annotated_genes %>%
  filter(Source=="VFDB"|Source=="UniProt_VF")

```


# Make a table of annotated vf gene differences between all hybrids and pangenome
```{r}
annotated_virulence_genes %>%
  select(sample,Locus,gene) %>%
  unique() %>%
  rowwise()%>%
  mutate(sample = ifelse(str_detect(sample,"hybrid"),analysis_design$Source[analysis_design$Sample==sample],sample),
         operon = ifelse(str_length(gene)==4,
                         substr(gene,1,3),
                         NA),
         present = 1) %>%
  pivot_wider(names_from = sample,values_from=present,values_fill=0) %>%
  relocate(operon,gene,Locus,pangenome) %>%
  arrange(gene)%>%
  write_tsv("../output/gene_analysis/VF_presence_table.tsv")
```

# Compare gene copies per strain, include pangenome. Save both a version with ubiquitus genes and a version with just varying genes
```{r}
# Make an overview of gene copies per strain per gene
gene_copies <- annotated_genes %>% filter(!isTRUE(Source=="UniProtKB" & is.na(gene) & is.na(Locus))) %>%
  group_by(Locus,sample)%>%
  mutate(copies=n(),
         gene = gsub("_.$","",gene)) %>%
  select(sample,gene,Locus,Source,copies) %>%
  distinct() %>%
  rowwise()%>%
  mutate(sample = ifelse(str_detect(sample,"hybrid"),analysis_design$Source[analysis_design$Sample==sample],sample))%>%
  ungroup()

# Write to tsv all copies
gene_copies %>%
  pivot_wider(names_from=sample,values_from=copies,values_fill=0)%>%
  relocate(gene,Locus,Source,pangenome) %>%
  write_tsv("../output/gene_analysis/gene_copies_per_strain.tsv")

# Remove genes present in all strains and pan genome
gene_copies %>%
  group_by(gene,Locus)%>%
  filter(n()!=9) %>%
  pivot_wider(names_from=sample,values_from=copies,values_fill=0)%>%
  relocate(gene,Locus,Source,pangenome) %>%
  arrange(pangenome,Source,gene) %>%
  write_tsv("../output/gene_analysis/varying_gene_copies_per_strain.tsv")
```


# Make a summary of all annotated genes per sample by category
```{r}
# Make a summary of all annotated genes per sample and save to file
annotation_counts_total <- annotated_genes %>%
  rowwise()%>%
  mutate(category = determine_annotation_category(type,product,Source),
         Locus = ifelse(Source=="UniProtKB" & is.na(Locus),gsub("_.$","",gene),Locus))%>%
  group_by(sample,category) %>%
  summarize(count=n()) %>% pivot_wider(names_from = category,values_from=count) %>%
  rowwise()%>%
  mutate(sample = ifelse(str_detect(sample,"hybrid"),analysis_design$Source[analysis_design$Sample==sample],sample))%>%
  arrange(sample)
  
# tsv file
annotation_counts_total %>%
  write_tsv("../output/gene_analysis/annotation_counts_total.tsv")
# delim file for latex table
annotation_counts_total %>%
  write_delim("../output/gene_analysis/annotation_counts_total_latex.txt",delim="&")

```
# Make a summary of all core-genome and all pan-genome genes
```{r}
# Make a summary of all core-genome and all pan-genome genes
overview_cg_pg <- annotated_genes %>%
  filter(!isTRUE(is.na(gene)&is.na(Locus)))%>%
  rowwise()%>%
  mutate(category = determine_annotation_category(type,product,Source),
         Locus = ifelse(Source=="UniProtKB" & is.na(Locus),gsub("_.$","",gene),Locus))%>%
  filter(sample != "pangenome") %>%
  select(sample,type,Locus,Source,product,category) %>%
  filter(type=="CDS" & category!="hypothetical protein") %>%
  select(-product)%>%
  distinct() %>%
  group_by(type,Locus)%>%
  mutate(presence = ifelse(n()==8,"core_genome","pan_genome")) %>%
  ungroup() %>%
  left_join(annotated_genes %>% select(gene,Locus),
            by="Locus")

# Calculate category counts for hybrid assemblies
category_counts_unique <- overview_cg_pg %>%
  select(Locus,category,presence) %>%
  distinct()%>%
  rowwise()%>%
  group_by(category,presence) %>%
  summarize(count=n()) %>%
  pivot_wider(names_from = presence,values_from=count)

# Calculate category counts for pangenome, add to the hybrid category counts and write to file
category_counts_unique <- annotated_genes %>%
  rowwise()%>%
  mutate(category = determine_annotation_category(type,product,Source),
         Locus = ifelse(Source=="UniProtKB" & is.na(Locus),gsub("_.$","",gene),Locus))%>%
  filter(sample == "pangenome")%>% 
  filter(type=="CDS" & category!="hypothetical protein") %>%
  select(-product)%>%
  select(Locus,category)%>%
  distinct() %>%
    group_by(category)%>%
  summarize(pangenome_assembly=n()) %>%
 right_join(category_counts_unique,by="category")

# Write as tsv file
category_counts_unique %>%
  write_tsv("../output/gene_analysis/category_counts_unique.tsv")
# Write as &-delimited latex file
category_counts_unique %>%
  write_delim("../output/gene_analysis/category_counts_unique_latex.txt","&")


```

# Compare pangenome and pan-genome
```{r}
pangenome_genes <-
  annotated_genes %>%
  filter(sample=="pangenome")%>%
  filter(!is.na(Locus)&type=="CDS")%>%
  pull("Locus")%>%
  unique()
hybrid_genes <-
  overview_cg_pg$Locus %>%
  unique()

pangenome_unique_genes <- pangenome_genes[!pangenome_genes %in%hybrid_genes]
hybrid_unique_genes <- hybrid_genes[!hybrid_genes %in% pangenome_genes]

annotated_genes %>%
  filter(Locus %in% pangenome_unique_genes) %>% 
  select(gene,Locus,Source)%>%
  distinct() %>%
  write_tsv("../output/gene_analysis/pangenome_unique_genes.tsv")
annotated_genes %>%
  filter(Locus %in% hybrid_unique_genes)%>% 
  select(gene,Locus,Source)%>%
  distinct()%>%
  write_tsv("../output/gene_analysis/hybrid_unique_genes.tsv")
```



# Make an overview of core-genome virulence factors which are highly similar in sequence and length
```{r}
# List of unique core-genome (cg) virulence factor (vf) genes
cg_vf_list <- overview_cg_pg %>% 
  filter(presence=="core_genome" & category=="Virulence Factor") %>% 
  pull(gene) %>% 
  unique()

# Make a list of cg vf genes which have the same protein sequence source and size in all genomes, and therefore are highly similar (hs).
hs_cg_vf_genes <- annotated_genes %>%
  filter(sample!="pangenome"& gene %in% cg_vf_list & type=="CDS")%>%
  select(sample,protein_id,sequence_length)%>%
  distinct()%>%
  group_by(protein_id,sequence_length)%>%
  filter(n()%%8==0) %>%
  select(-sample)%>%
  distinct() %>%
  mutate(id_length = paste0(protein_id,"_",sequence_length)) %>%
  pull("id_length")

# Make an overview of all annotations of these genes and save as a tsv file
annotated_genes %>%
  filter(paste0(protein_id,"_",sequence_length) %in% hs_cg_vf_genes & sample != "pangenome") %>%
  write_tsv("../output/gene_analysis/highly_similar_cg_vf_genes.tsv")
```

```{r}
# Make a duplicate mtase dataframe, starting with all genes which are duplicates of genes that have been annotated. 
# The existence of these duplicates is likely the cause of genes being discovered in different organisms.
mtase_genes_df <- 
  d_pdb %>% 
  # Filter these duplicates for genes of which the duplicate has been annotated and which is sourced from the REBASE gold standard
  filter(annotated %in% annotated_genes$gene & Source=="REBASE") %>%
  # For each of these genes, add the samples in which the duplicate MTase was found.
  left_join(annotated_genes %>% select(sample,gene,start,end,contig),
            by=c("annotated"="gene")) %>%
  mutate(gene=Gene)%>%
  # Remove colums
  select(-c(Locus,annotated,sequence)) %>%
  # Add mtase genes which were annotated in the genomes
  full_join(annotated_genes %>% filter(Source=="REBASE") %>% select(sample,gene,Source,product,start,end,contig),
            by=c("sample","gene","Source","metadata"="product","start","end","contig")) %>%
  # Make a new column for the motif
  mutate(motif = gsub("recognition_sequence: *","",metadata)) %>%
  # Remove excess columns
  select(-c(Gene,metadata))%>%
  # filter NA values (formatted as "NA" by VFDB)
  filter(motif!="NA") %>%
  rowwise()%>%
  # add serovar
  select(-Source)%>%
  left_join(analysis_design %>% select(Sample,Source),
            by=c("sample"="Sample"))%>%
  select(Source,contig,start,motif,gene)%>%
  filter(!is.na(Source))

# Make an overview of, per CDS (or site), how many RM elements are found related to a motif and list one of them.
rm_element_overview <- mtase_genes_df %>%
  select(Source,contig,start)%>%
  distinct()%>%
  group_by(Source)%>%
  arrange(contig,start)%>%
  mutate(site = paste0("site ",row_number())) %>%
  right_join(mtase_genes_df,
             by=c("Source","contig","start")) %>%
  group_by(Source,site,start,motif)%>%
  summarize(`Homologous RM elements` = n()) %>%
  ungroup()%>%
  left_join(mtase_genes_df %>% 
              group_by(Source,contig,start,motif) %>%
              arrange(gene)%>%
              filter(row_number()==1),
             by=c("Source","start","motif")) %>% 
  rename(gene="Putative RM element",
         motif="Motif",
         contig="Contig",
         site="Site",
         Source="Sample",
         start="Start")%>%
  select(Sample,Site,Motif,`Homologous RM elements`,`Putative RM element`) %>%
  arrange(Sample,Site)

# Write to TSV file
rm_element_overview %>%
  write_tsv("../output/gene_analysis/rm_overview_latex_table.tsv")

# Write to .tex file
rm_element_overview %>% 
  kable(format="latex",
        vline="",
        booktabs=TRUE,
        longtable=TRUE) %>%
  collapse_rows(columns=c(1:4),
                valign="top",
                longtable_clean_cut = FALSE) %>%
  cat(file="../output/gene_analysis/rm_overview_latex_table.tex",
      sep="\n")
```

